network/netcore/src/transport/tcp.rs

```diff
@@ -93,14 +94,17 @@ impl Transport for TcpTransport {
     }
 
     fn dial(&self, addr: NetworkAddress) -> Result<Self::Outbound, Self::Error> {
-        let (addr_string, _addr_suffix) =
-            parse_addr_and_port(addr.as_slice()).ok_or_else(|| invalid_addr_error(&addr))?;
+        let protos = addr.as_slice();
+
+        // ensure addr is well formed to save some work before potentially
+        // spawning a dial task that will fail anyway.
         // TODO(philiphayes): base tcp transport should not allow trailing protocols
-        // TODO(philiphayes): use `tokio::net::lookup_host()` then filter the results
-        // so e.g. `Protocol::Dns4` only returns `SocketAddrV4`s and `Protocol::Dns6`
-        // only returns `SocketAddrV6`.
+        None.or_else(|| parse_ip_tcp(protos).map(|_| ()))
+            .or_else(|| parse_dns_tcp(protos).map(|_| ()))
+            .ok_or_else(|| invalid_addr_error(&addr))?;
```

Comment:
I'm not sure what the`None.or_else(` paradigm does.  But I do get that you're parsing for IP/TCP & DNS/TCP to see if the address is valid.

Reply:
can't it be just 
```
parse_ip_tcp(protos).map(|_| ()).or_else(parse_dns_tcp(protos).map(|_| ()).ok_or_else(..)?;
```

Reply:
the only reason for the `None.or_else` (which is effectively a no-op) is so the parsing lines align. cargo fmt would otherwise align this block like

```rust
parse_ip_tcp(protos)
    .map(|_| ())
    .or_else(|| parse_dns_tcp(protos).map(|_| ()))
    .ok_or_else(|| invalid_addr_error(&addr))?;
```

kind of on-the-fence about this pattern... I think it looks nicer but is also clearly more confusing : (

@zekun000 what would you prefer?

Reply:
I'd go over straightforward code than looking nicer.  That way the next guy doesn't come in and go, what is this `None.or_else()` and ask the same question.

Reply:
agree, I prefer straightforward and less confusing one too

Reply:
done!